src/back/spv/block.rs

```diff
@@ -696,6 +696,88 @@ impl<'w> BlockContext<'w> {
+                                self.temp_list.clear();
+                                self.temp_list.resize(
+                                    size as _,
+                                    self.writer
+                                        .get_constant_scalar(crate::ScalarValue::Float(0.0), width),
+                                );
+
+                                let id = self.gen_id();
+                                block.body.push(Instruction::composite_construct(
+                                    result_type_id,
+                                    id,
+                                    &self.temp_list,
+                                ));
+
+                                id
+                            };
+
+                            let arg2_id = {
+                                self.temp_list.clear();
+                                self.temp_list.resize(
+                                    size as _,
+                                    self.writer
+                                        .get_constant_scalar(crate::ScalarValue::Float(1.0), width),
+                                );
+
+                                let id = self.gen_id();
+                                block.body.push(Instruction::composite_construct(
+                                    result_type_id,
+                                    id,
+                                    &self.temp_list,
+                                ));
+
+                                id
+                            };
+
+                            MathOp::Custom(Instruction::ext_inst(
+                                self.writer.gl450_ext_inst_id,
+                                spirv::GLOp::FClamp,
+                                result_type_id,
+                                id,
+                                &[arg0_id, arg1_id, arg2_id],
+                            ))
+                        }
+                        crate::TypeInner::Scalar {
+                            kind: crate::ScalarKind::Float,
+                            width,
+                        } => {
+                            let arg1_id = self
+                                .writer
+                                .get_constant_scalar(crate::ScalarValue::Float(0.0), width);
+                            let arg2_id = self
+                                .writer
+                                .get_constant_scalar(crate::ScalarValue::Float(1.0), width);
+
+                            MathOp::Custom(Instruction::ext_inst(
+                                self.writer.gl450_ext_inst_id,
+                                spirv::GLOp::FClamp,
+                                result_type_id,
+                                id,
+                                &[arg0_id, arg1_id, arg2_id],
+                            ))
+                        }
+                        ref other => unimplemented!("Unexpected saturate({:?})", other),
+                    },
```

Comment:
This could could be greatly simplified, we could do something like this
```suggestion
                    Mf::Saturate => {
                      let (maybe_size, width) = match *arg_ty {
                          crate::TypeInner::Vector {
                              size,
                              width,
                              ..
                          } => (Some(size), width),
                          crate::TypeInner::Scalar { width, .. } => (None, width),
                          ref other => unimplemented!("Unexpected saturate({:?})", other),
                      };
                      
                      let mut arg1_id = self
                          .writer
                          .get_constant_scalar(crate::ScalarValue::Float(0.0), width);
                      let mut arg2_id = self
                          .writer
                          .get_constant_scalar(crate::ScalarValue::Float(1.0), width);
                          
                      if let Some(size) = maybe_size {
                        let value = LocalType::Value {
                            vector_size: Some(size),
                            kind: crate::ScalarKind::Float,
                            width,
                            pointer_space: None,
                        };

                        let result_type_id = self.get_type_id(LookupType::Local(value));
  
                        self.temp_list.clear();
                        self.temp_list.resize(size as _, arg1_id);

                        let id = self.gen_id();
                        block.body.push(Instruction::composite_construct(
                            result_type_id,
                            id,
                            &self.temp_list,
                        ));
                        arg1_id = id;
  
                        self.temp_list.clear();
                        self.temp_list.resize(size as _, arg2_id);

                        let id = self.gen_id();
                        block.body.push(Instruction::composite_construct(
                            result_type_id,
                            id,
                            &self.temp_list,
                        ));
                        arg2_id = id;
                      }

                      MathOp::Custom(Instruction::ext_inst(
                          self.writer.gl450_ext_inst_id,
                          spirv::GLOp::FClamp,
                          result_type_id,
                          id,
                          &[arg0_id, arg1_id, arg2_id],
                      ))
                   },
```